equinix/resource_metal_project.go

```diff
@@ -1,25 +1,28 @@
 package equinix
 
 import (
+	"context"
 	"fmt"
 	"path"
 	"strings"
+	"time"
 
+	"github.com/equinix/equinix-sdk-go/services/metalv1"
 	equinix_errors "github.com/equinix/terraform-provider-equinix/internal/errors"
 
 	"github.com/equinix/terraform-provider-equinix/internal/config"
 
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
-	"github.com/packethost/packngo"
 )
 
 func resourceMetalProject() *schema.Resource {
 	return &schema.Resource{
-		Create: resourceMetalProjectCreate,
-		Read:   resourceMetalProjectRead,
-		Update: resourceMetalProjectUpdate,
-		Delete: resourceMetalProjectDelete,
+		CreateContext: resourceMetalProjectCreate,
```

Comment:
`CreateContext` will by default introduce a default timeout of 20 minutes
https://developer.hashicorp.com/terraform/plugin/sdkv2/resources/retries-and-customizable-timeouts#default-timeouts-and-deadline-exceeded-errors

We will have to use `CreateWithoutTimeout` where no officially timeout is configured

and same for others

Reply:
It's not clear to me when we should or shouldn't use the functions that can time out.  CRUD operations on a project should generally be very fast, so most of the time it probably won't matter if we use the `...Context` or `...WithoutTimeout` functions.  If, for some reason, it took longer than 20 minutes, I think it would be preferable for the resource to fail than to hang indefinitely.

Reply:
Yes Ideally thats true. But I guess some resources may take more time and users would have to provide that explicitly via timeout block. AFAIK we have [officially documented timeouts](https://github.com/equinix/terraform-provider-equinix/blob/main/docs/resources/equinix_metal_device.md#timeouts) for certain resources and same is being configured in the code [as here](https://github.com/equinix/terraform-provider-equinix/blob/90516e3aa71d94696753723bc6a5a20e685fdc88/equinix/resource_metal_device.go#L43C3-L43C37)

For those places we can have `...Context` and wherever we did not have any timeout configured there we can have `...WithoutTimeout` because we have not declared timeouts so a resource may take more than 20 mins for the user and we are not interrupting it currently

Reply:
I'm skeptical of the idea that we should let fast actions have the risk of hanging forever simply because we haven't established the exact definition of "fast," but at least 99% of the time the choice of function won't matter for this resource.   In light of that, I've updated the CRUD functions to use the `WithoutTimeout` versions.

Reply:
Yes completely agree with you. Ideally any cloud resource (which is costing user) should have a timeout with the provision to autocleanup failed, unknown status resources.